clang/lib/Frontend/InitPreprocessor.cpp

```diff
@@ -671,10 +671,8 @@ static void InitializeCPlusPlusFeatureTestMacros(const LangOptions &LangOpts,
                         LangOpts.CPlusPlus23   ? "202211L"
                         : LangOpts.CPlusPlus17 ? "201603L"
                                                : "200907");
-    Builder.defineMacro("__cpp_static_assert", LangOpts.CPlusPlus26 ? "202306L"
-                                               : LangOpts.CPlusPlus17
-                                                   ? "201411L"
-                                                   : "200410");
+    // C++17 / C++26 static_assert backported
+    Builder.defineMacro("__cpp_static_assert", "202306L");
```

Comment:
AFAICT this is non-conforming. C++20 requires `__cpp_static_assert` to be `201411L`. Personally I think that's a defect in the standard, but that's not my decision to make. Also note that other feature test macros reflect the standards wording as well and not whether the feature has been backported.

Reply:
I see that other features that have been backported only define the feature test macro if it didn't exist in the older versions. This sounds reasonable, I'll revert this change.

Reply:
I'm not certain I agree that this is non-conforming, but the standard could be made more clear. The *name* of the feature test macro has to be predefined, but the standard says nothing about the value beyond what the value is for that particular release of the standard. The whole point to feature test macros is to be able to use the resulting date to tell you which feature you're working with, so not being able to specify an updated value for backported functionality... kind of defeats the purpose.

Reply:
I'm not exactly sure what's ambiguous here. You say yourself that it specifies what the value should be for a given standards version. I also can't find anything suggesting that implementations are allowed to define then to something else. Anyways, since we seem to agree that implementations should be allowed to define them to some higher number, I'm not sure it makes a ton of sense to do anything other than filing an issue. But where should it be filed? Is that core wording or library? Both? Neither?

Reply:
Oh, also note that this could potentially be a huge breaking change, since users may rely on them guarding for a version.

Reply:
> I'm not exactly sure what's ambiguous here. You say yourself that it specifies what the value should be for a given standards version.

https://eel.is/c++draft/cpp.predefined has two styles of predefined macros. Ones where there is normative text explaining what the macro has to expand to, and others where it simply lists a value in a table. Nothing in that subclause says "shall be defined as" for the values in the table. It just says the *names* shall be defined.

> Anyways, since we seem to agree that implementations should be allowed to define them to some higher number, I'm not sure it makes a ton of sense to do anything other than filing an issue. But where should it be filed? Is that core wording or library? Both? Neither?

Probably one Core issue and one Library issue.

Worth remembering that all of these came out of [SD6](https://isocpp.org/std/standing-documents/sd-6-sg10-feature-test-recommendations) which says: "The value specified for a feature-test macro is based on the year and month in which the feature is voted into the working draft. In a case where a feature is subsequently changed in a significant way, but arguably remains the same feature, the value of the macro is changed to indicate the “revision level” of the specification of the feature. However, in most cases it is expected that the presence of a feature can be determined by the presence of any non-zero macro value; for example:"

> Oh, also note that this could potentially be a huge breaking change, since users may rely on them guarding for a version.

Huh? The idiomatic use is `#if defined(__cplusplus) && __cpp_whatever >= 201907L`, so bumping to a later revision is not a breaking change. It's of course possible for a user to test using `==` or some other approach, but they're using the feature testing macro incorrectly at that point.

Reply:
Looking at how we handle other backported features, it appears that we didn't change anything about feature test macros when we enabled them. In other words, what was undefined remain undefined, what was defined is still defined to have the same value as if backport never happened.

Reply:
I suggest that we remain consistent with that

Reply:
> https://eel.is/c++draft/cpp.predefined has two styles of predefined macros. Ones where there is normative text explaining what the macro has to expand to, and others where it simply lists a value in a table. Nothing in that subclause says "shall be defined as" for the values in the table. It just says the _names_ shall be defined.

Hmm, true. The library version seems less ambiguous, since there it's in the synopsis as `#define __cpp_lib_whatever <some_number>L`.

> Probably one Core issue and one Library issue.

Sounds good. Who should file them?

> Huh? The idiomatic use is `#if defined(__cplusplus) && __cpp_whatever >= 201907L`, so bumping to a later revision is not a breaking change. It's of course possible for a user to test using `==` or some other approach, but they're using the feature testing macro incorrectly at that point.

I'm more thinking of stuff like this: https://godbolt.org/z/qzaMaP4sW.

Reply:
> I'm more thinking of stuff like this: https://godbolt.org/z/qzaMaP4sW.

That problem already exists for these macros:

https://github.com/llvm/llvm-project/blob/4f067dc467a1f97a1b6ac420926da8bab5f243bf/clang/lib/Frontend/InitPreprocessor.cpp#L758-L767

https://godbolt.org/z/nGqPb4Pvr

Reply:
We have had that discussion before. https://github.com/llvm/llvm-project/issues/77262

I think we should
 * unconditionally set the value of __cpp_static_assert to 202306L in this patch
 * Explore through an rfc or more detailed issue an holistic change of there is consensus that we should improve how pedantic errors and feature test macro interact.

Reply:
@MitalAshok That doesn't mean that it won't break people though. Maybe not in this specific case, but almost certainly in general.

Reply:
I agree with @cor3ntin -- pedantic errors and feature testing are a wider concern; in addition to #77262 we also have #84372 which is a similar problem but with `__has_extension` and pedantic errors.

In your example, the user is getting precisely what they asked for and I don't see the issue. They explicitly opt in to diagnostics about code that uses extensions in C++14 mode, so the diagnostic warning is expected. And they explicitly opt to make all warnings be errors, so upgrading the warning to an error is expected.